### ISSUE TITLE
Memory leak when logging validation metrics outside of validation_step

### ISSUE THREAD

**Issue by ido-greenfeld** [PyTorchLightning/pytorch-lightning#8215]
## 🐛 Bug

I was facing a weird memory issue.
I train a segmentation model with two different configurations.

The first -
- Image size 256X256
- Batch size 120
- 4 GPUs
The second -
- Image size 512X512
- Batch size 16
- 4 GPUs

The first configuration runs flawlessly. I run it for 30 epochs and the memory on each GPU remains constant.
On the second configuration however, I manage to run the training stage with constant memory but when I reach the validation stage something goes wrong.
The validation stage has ~200 steps. I go through the first ~100 without an issue, memory stays constant! but after that all of a sudden I get a spike in all 4 GPUs memory and OOM error.

That memory spike doesn't happen on the first configuration. The number of validation step in the first configuration is ~40...
So I tried recreating this on the smaller images to see if somehow the image size has something to do with the issue.
I duplicated my validation set so that now I have ~120 steps in the validation stage and lo and behold I now get the same issue after ~80 steps!

So for some reason if the validation stage has more than X steps I get this weird memory spike which is not gradual, it happens all of a sudden in the middle of the validation stage.
Maybe after X validation steps the dataloaders start loading new training data for the next epoch for some reason?

I was able to fix it by doing the following steps - 
In my original code I had the following functions inside my LightningModule

```
    def predict(self, x):
        y_hat = self(x)
        preds_classified = torch.argmax(preds, dim=1)
        return preds_classified, y_hat
        
    def training_step(self, batch, batch_idx):
        x, y = batch
        preds_classified, y_hat = self.predict(x)
        loss = self.loss_fn(y_hat, y)
        return {'loss': loss, 'preds': preds_classified, 'target': y}
        
    def validation_step(self, batch, batch_idx):
        results = self.training_step(batch, batch_idx)
        return results
        
    def validation_step_end(self, outputs):
        self.log('iou', {'val_iou': self.validation_iou(outputs['preds'], outputs['target'])}, on_step=False,
                 on_epoch=True)
        return outputs
        
    def validation_epoch_end(self, outputs):
        if self.sample_val is not None:
            self.visualize_segmentation(self.sample_val, 'val_images_sample_preds')
        losses = [output['loss'] for output in outputs]
        self.log('loss', {'val_loss': torch.stack(losses, dim=0).sum(dim=0).sum(dim=0).detach() / len(outputs)})
```

So I had two values which I logged `val_iou ` and `val_loss` and they both came from `training_step` and were sent as output from `validation_step` to be used in `validation_step_end` and `validation_epoch_end`
Not the most clean implementation... but it worked...

After diving into it a bit more, I noticed that if I don't output  the full output from validation_step but only the 'loss': loss part (and of course comment out the lines that used the other parameters) I don't get the memory spike anymore...
So I changed my code and I now log the parameters inside `validation_step` like this -

```
    def validation_step(self, batch, batch_idx):
        results = self.training_step(batch, batch_idx)
        self.log('iou', {'val_iou': self.validation_iou(results['preds'], results['target'])}, on_step=False, on_epoch=True)
        self.log('loss', {'val_loss': results['loss']}, on_step=False, on_epoch=True)
        return {'loss': results['loss']}
```

This seems to give the same results without the memory spike... I now get constant memory throughout the train and validations stages...
What I really don't understand is why did the memory start to build up only after ~100 steps... this happened again and again under different images and batch sizes... always in the range of 80-120 steps

Unfortunately I was unable to recreate this issue with BoringModel

### Expected behavior

Memory should remain constant throughout validation stage.

### Environment

 - PyTorch Version (e.g., 1.0): 1.8.1
 - OS (e.g., Linux): Ubuntu 20.04 LTS
 - How you installed PyTorch (`conda`, `pip`, source): PIP
 - Build command you used (if compiling from source): pip install pytorch-lightning
 - Python version: 3.8.4
 - CUDA/cuDNN version: 11.1
 - GPU models and configuration: 4X Nvidia T4

**Comment by ethanwharris** [comment]
@ido-greenfeld thanks for the issue! Are you able to provide some example code that caused the error in addition to the example code that worked? Can I ask how you are tracking the validation progress? Since we track both training steps and validation steps on a single progress bar, is it possible that the spike is actually occuring at the beginning of validation?

**Comment by ido-greenfeld** [comment]
hi @ethanwharris ,
This is my lightningModule with the issue - 

```
import pytorch_lightning as pl
import segmentation_models_pytorch as smp
import torch
import torchmetrics
from torch import nn, optim


class UnetLightning(pl.LightningModule):
    def __init__(self, num_classes=2, encoder_name='resnet34', encoder_weights='imagenet', lr=0.001, in_channels=3):
        super().__init__()
        self.unet_model = smp.Unet(
            encoder_name=encoder_name,  # choose encoder, e.g. mobilenet_v2 or efficientnet-b7
            encoder_weights=encoder_weights,  # use `imagenet` pre-trained weights for encoder initialization
            in_channels=in_channels,  # model input channels (1 for gray-scale images, 3 for RGB, etc.)
            classes=num_classes,  # model output channels (number of classes in your dataset)
        )

        self.train_iou = torchmetrics.IoU(num_classes=num_classes, ignore_index=0)
        self.validation_iou = torchmetrics.IoU(num_classes=num_classes, ignore_index=0)

        self.loss_fn = nn.CrossEntropyLoss()
        self.lr = lr

        self.save_hyperparameters()

    def freeze(self):
        for param in self.unet_model.encoder.parameters():
            param.requires_grad = False

    def configure_optimizers(self):
        self.lr = self.lr * self.trainer.gpus if self.trainer.gpus > 1 else self.lr
        return optim.Adam([dict(params=self.parameters(), lr=self.lr)])

    def forward(self, x):
        y_hat = self.unet_model(x)
        return y_hat

    def predict(self, x):
        y_hat = self(x)
        preds_classified = torch.argmax(y_hat, dim=1)
        return preds_classified, y_hat

    def training_step(self, batch, batch_idx):
        x, y = batch
        preds_classified, y_hat = self.predict(x)
        loss = self.loss_fn(y_hat, y)

        return {'loss': loss, 'preds': preds_classified, 'target': y}

    def training_step_end(self, outputs):
        self.log('loss', {'train_loss': outputs['loss'].detach()}, on_step=True, on_epoch=False)
        self.log('iou', {'train_iou': self.train_iou(outputs['preds'], outputs['target'])}, on_step=True,
                 on_epoch=False)

        return outputs

    def validation_step(self, batch, batch_idx):
        results = self.training_step(batch, batch_idx)

        return results

    def validation_step_end(self, outputs):
        self.log('iou', {'val_iou': self.validation_iou(outputs['preds'], outputs['target'])}, on_step=False,
                 on_epoch=True)
        return outputs

    def validation_epoch_end(self, outputs):
        losses = [output['loss'] for output in outputs]
        self.log('loss', {'val_loss': torch.stack(losses, dim=0).sum(dim=0).sum(dim=0).detach() / len(outputs)})
```

And this is the code without the issue - 

```
import pytorch_lightning as pl
import segmentation_models_pytorch as smp
import torch
import torchmetrics
from torch import nn, optim


class UnetLightning(pl.LightningModule):
    def __init__(self, num_classes=2, encoder_name='resnet34', encoder_weights='imagenet', lr=0.001, in_channels=3):
        super().__init__()
        self.unet_model = smp.Unet(
            encoder_name=encoder_name,  # choose encoder, e.g. mobilenet_v2 or efficientnet-b7
            encoder_weights=encoder_weights,  # use `imagenet` pre-trained weights for encoder initialization
            in_channels=in_channels,  # model input channels (1 for gray-scale images, 3 for RGB, etc.)
            classes=num_classes,  # model output channels (number of classes in your dataset)
        )

        self.train_iou = torchmetrics.IoU(num_classes=num_classes, ignore_index=0)
        self.validation_iou = torchmetrics.IoU(num_classes=num_classes, ignore_index=0)

        self.loss_fn = nn.CrossEntropyLoss()
        self.lr = lr

        self.save_hyperparameters()

    def freeze(self):
        for param in self.unet_model.encoder.parameters():
            param.requires_grad = False

    def configure_optimizers(self):
        self.lr = self.lr * self.trainer.gpus if self.trainer.gpus > 1 else self.lr
        return optim.Adam([dict(params=self.parameters(), lr=self.lr)])

    def forward(self, x):
        y_hat = self.unet_model(x)
        return y_hat

    def predict(self, x):
        y_hat = self(x)
        preds_classified = torch.argmax(y_hat, dim=1)
        return preds_classified, y_hat

    def training_step(self, batch, batch_idx):
        x, y = batch
        preds_classified, y_hat = self.predict(x)
        loss = self.loss_fn(y_hat, y)

        return {'loss': loss, 'preds': preds_classified, 'target': y}

    def training_step_end(self, outputs):
        self.log('loss', {'train_loss': outputs['loss'].detach()}, on_step=True, on_epoch=False)
        self.log('iou', {'train_iou': self.train_iou(outputs['preds'], outputs['target'])}, on_step=True,
                 on_epoch=False)

        return outputs

    def validation_step(self, batch, batch_idx):
        results = self.training_step(batch, batch_idx)

        self.log('iou', {'val_iou': self.validation_iou(results['preds'], results['target'])}, on_step=False, on_epoch=True)
        self.log('loss', {'val_loss': results['loss']}, on_step=False, on_epoch=True)

        return {'loss': results['loss']}
```

Notice the changes of where I log the metrics in the validation stage.
My data are segmentation images and masks.

As for how I'm tracking the validation progress - I was debugging the progress in PyCharm, I managed to find the appx validation step where this issue happens and set a breakpoint inside `validation_step`. By going step by step I narrowed it down to the logging issue.
I also verified that `model.training` is False at the breakpoint where this happens.

Regarding the single progress bar - this is weird, I see 2 progress bars. One for training + validation and one for validation (which appears only in the validation stage and then disappears)

**Comment by ethanwharris** [comment]
Hi @ido-greenfeld thanks for providing the scripts :smiley: 

> Regarding the single progress bar - this is weird, I see 2 progress bars. One for training + validation and one for validation (which appears only in the validation stage and then disappears)

Yes, sorry, that's correct

Can I ask what pytorch-lightning version you are using? There were some recent improvements around logging and memory consumption, could you check that the issue is still there when installing from master?
```
pip install https://github.com/PyTorchLightning/pytorch-lightning/archive/master.zip
```

**Comment by ido-greenfeld** [comment]
Hi @ethanwharris sorry for the late reply
I'm using pytorch-lightning version 1.3.5
I'll try and recreate the issue on master but it might take some time
